// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: api/manager/feedback/manager_feedback_category.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ListFeedbackCategoryRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListFeedbackCategoryRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListFeedbackCategoryRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListFeedbackCategoryRequestMultiError, or nil if none found.
func (m *ListFeedbackCategoryRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListFeedbackCategoryRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetPage() < 1 {
		err := ListFeedbackCategoryRequestValidationError{
			field:  "Page",
			reason: "value must be greater than or equal to 1",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if val := m.GetPageSize(); val < 1 || val > 50 {
		err := ListFeedbackCategoryRequestValidationError{
			field:  "PageSize",
			reason: "value must be inside range [1, 50]",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.Order != nil {

		if _, ok := _ListFeedbackCategoryRequest_Order_InLookup[m.GetOrder()]; !ok {
			err := ListFeedbackCategoryRequestValidationError{
				field:  "Order",
				reason: "value must be in list [asc desc]",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if m.OrderBy != nil {

		if _, ok := _ListFeedbackCategoryRequest_OrderBy_InLookup[m.GetOrderBy()]; !ok {
			err := ListFeedbackCategoryRequestValidationError{
				field:  "OrderBy",
				reason: "value must be in list [id]",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if m.Name != nil {
		// no validation rules for Name
	}

	if len(errors) > 0 {
		return ListFeedbackCategoryRequestMultiError(errors)
	}

	return nil
}

// ListFeedbackCategoryRequestMultiError is an error wrapping multiple
// validation errors returned by ListFeedbackCategoryRequest.ValidateAll() if
// the designated constraints aren't met.
type ListFeedbackCategoryRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListFeedbackCategoryRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListFeedbackCategoryRequestMultiError) AllErrors() []error { return m }

// ListFeedbackCategoryRequestValidationError is the validation error returned
// by ListFeedbackCategoryRequest.Validate if the designated constraints
// aren't met.
type ListFeedbackCategoryRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListFeedbackCategoryRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListFeedbackCategoryRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListFeedbackCategoryRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListFeedbackCategoryRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListFeedbackCategoryRequestValidationError) ErrorName() string {
	return "ListFeedbackCategoryRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListFeedbackCategoryRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListFeedbackCategoryRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListFeedbackCategoryRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListFeedbackCategoryRequestValidationError{}

var _ListFeedbackCategoryRequest_Order_InLookup = map[string]struct{}{
	"asc":  {},
	"desc": {},
}

var _ListFeedbackCategoryRequest_OrderBy_InLookup = map[string]struct{}{
	"id": {},
}

// Validate checks the field values on ListFeedbackCategoryReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListFeedbackCategoryReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListFeedbackCategoryReply with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListFeedbackCategoryReplyMultiError, or nil if none found.
func (m *ListFeedbackCategoryReply) ValidateAll() error {
	return m.validate(true)
}

func (m *ListFeedbackCategoryReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Total

	for idx, item := range m.GetList() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListFeedbackCategoryReplyValidationError{
						field:  fmt.Sprintf("List[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListFeedbackCategoryReplyValidationError{
						field:  fmt.Sprintf("List[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListFeedbackCategoryReplyValidationError{
					field:  fmt.Sprintf("List[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListFeedbackCategoryReplyMultiError(errors)
	}

	return nil
}

// ListFeedbackCategoryReplyMultiError is an error wrapping multiple validation
// errors returned by ListFeedbackCategoryReply.ValidateAll() if the
// designated constraints aren't met.
type ListFeedbackCategoryReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListFeedbackCategoryReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListFeedbackCategoryReplyMultiError) AllErrors() []error { return m }

// ListFeedbackCategoryReplyValidationError is the validation error returned by
// ListFeedbackCategoryReply.Validate if the designated constraints aren't met.
type ListFeedbackCategoryReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListFeedbackCategoryReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListFeedbackCategoryReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListFeedbackCategoryReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListFeedbackCategoryReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListFeedbackCategoryReplyValidationError) ErrorName() string {
	return "ListFeedbackCategoryReplyValidationError"
}

// Error satisfies the builtin error interface
func (e ListFeedbackCategoryReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListFeedbackCategoryReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListFeedbackCategoryReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListFeedbackCategoryReplyValidationError{}

// Validate checks the field values on CreateFeedbackCategoryRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateFeedbackCategoryRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateFeedbackCategoryRequest with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// CreateFeedbackCategoryRequestMultiError, or nil if none found.
func (m *CreateFeedbackCategoryRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateFeedbackCategoryRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetName()) < 1 {
		err := CreateFeedbackCategoryRequestValidationError{
			field:  "Name",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return CreateFeedbackCategoryRequestMultiError(errors)
	}

	return nil
}

// CreateFeedbackCategoryRequestMultiError is an error wrapping multiple
// validation errors returned by CreateFeedbackCategoryRequest.ValidateAll()
// if the designated constraints aren't met.
type CreateFeedbackCategoryRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateFeedbackCategoryRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateFeedbackCategoryRequestMultiError) AllErrors() []error { return m }

// CreateFeedbackCategoryRequestValidationError is the validation error
// returned by CreateFeedbackCategoryRequest.Validate if the designated
// constraints aren't met.
type CreateFeedbackCategoryRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateFeedbackCategoryRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateFeedbackCategoryRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateFeedbackCategoryRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateFeedbackCategoryRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateFeedbackCategoryRequestValidationError) ErrorName() string {
	return "CreateFeedbackCategoryRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreateFeedbackCategoryRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateFeedbackCategoryRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateFeedbackCategoryRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateFeedbackCategoryRequestValidationError{}

// Validate checks the field values on CreateFeedbackCategoryReply with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateFeedbackCategoryReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateFeedbackCategoryReply with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateFeedbackCategoryReplyMultiError, or nil if none found.
func (m *CreateFeedbackCategoryReply) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateFeedbackCategoryReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return CreateFeedbackCategoryReplyMultiError(errors)
	}

	return nil
}

// CreateFeedbackCategoryReplyMultiError is an error wrapping multiple
// validation errors returned by CreateFeedbackCategoryReply.ValidateAll() if
// the designated constraints aren't met.
type CreateFeedbackCategoryReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateFeedbackCategoryReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateFeedbackCategoryReplyMultiError) AllErrors() []error { return m }

// CreateFeedbackCategoryReplyValidationError is the validation error returned
// by CreateFeedbackCategoryReply.Validate if the designated constraints
// aren't met.
type CreateFeedbackCategoryReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateFeedbackCategoryReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateFeedbackCategoryReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateFeedbackCategoryReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateFeedbackCategoryReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateFeedbackCategoryReplyValidationError) ErrorName() string {
	return "CreateFeedbackCategoryReplyValidationError"
}

// Error satisfies the builtin error interface
func (e CreateFeedbackCategoryReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateFeedbackCategoryReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateFeedbackCategoryReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateFeedbackCategoryReplyValidationError{}

// Validate checks the field values on UpdateFeedbackCategoryRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateFeedbackCategoryRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateFeedbackCategoryRequest with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// UpdateFeedbackCategoryRequestMultiError, or nil if none found.
func (m *UpdateFeedbackCategoryRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateFeedbackCategoryRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() < 1 {
		err := UpdateFeedbackCategoryRequestValidationError{
			field:  "Id",
			reason: "value must be greater than or equal to 1",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetName()) < 1 {
		err := UpdateFeedbackCategoryRequestValidationError{
			field:  "Name",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return UpdateFeedbackCategoryRequestMultiError(errors)
	}

	return nil
}

// UpdateFeedbackCategoryRequestMultiError is an error wrapping multiple
// validation errors returned by UpdateFeedbackCategoryRequest.ValidateAll()
// if the designated constraints aren't met.
type UpdateFeedbackCategoryRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateFeedbackCategoryRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateFeedbackCategoryRequestMultiError) AllErrors() []error { return m }

// UpdateFeedbackCategoryRequestValidationError is the validation error
// returned by UpdateFeedbackCategoryRequest.Validate if the designated
// constraints aren't met.
type UpdateFeedbackCategoryRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateFeedbackCategoryRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateFeedbackCategoryRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateFeedbackCategoryRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateFeedbackCategoryRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateFeedbackCategoryRequestValidationError) ErrorName() string {
	return "UpdateFeedbackCategoryRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateFeedbackCategoryRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateFeedbackCategoryRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateFeedbackCategoryRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateFeedbackCategoryRequestValidationError{}

// Validate checks the field values on UpdateFeedbackCategoryReply with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateFeedbackCategoryReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateFeedbackCategoryReply with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateFeedbackCategoryReplyMultiError, or nil if none found.
func (m *UpdateFeedbackCategoryReply) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateFeedbackCategoryReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return UpdateFeedbackCategoryReplyMultiError(errors)
	}

	return nil
}

// UpdateFeedbackCategoryReplyMultiError is an error wrapping multiple
// validation errors returned by UpdateFeedbackCategoryReply.ValidateAll() if
// the designated constraints aren't met.
type UpdateFeedbackCategoryReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateFeedbackCategoryReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateFeedbackCategoryReplyMultiError) AllErrors() []error { return m }

// UpdateFeedbackCategoryReplyValidationError is the validation error returned
// by UpdateFeedbackCategoryReply.Validate if the designated constraints
// aren't met.
type UpdateFeedbackCategoryReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateFeedbackCategoryReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateFeedbackCategoryReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateFeedbackCategoryReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateFeedbackCategoryReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateFeedbackCategoryReplyValidationError) ErrorName() string {
	return "UpdateFeedbackCategoryReplyValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateFeedbackCategoryReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateFeedbackCategoryReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateFeedbackCategoryReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateFeedbackCategoryReplyValidationError{}

// Validate checks the field values on DeleteFeedbackCategoryRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteFeedbackCategoryRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteFeedbackCategoryRequest with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// DeleteFeedbackCategoryRequestMultiError, or nil if none found.
func (m *DeleteFeedbackCategoryRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteFeedbackCategoryRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() < 1 {
		err := DeleteFeedbackCategoryRequestValidationError{
			field:  "Id",
			reason: "value must be greater than or equal to 1",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return DeleteFeedbackCategoryRequestMultiError(errors)
	}

	return nil
}

// DeleteFeedbackCategoryRequestMultiError is an error wrapping multiple
// validation errors returned by DeleteFeedbackCategoryRequest.ValidateAll()
// if the designated constraints aren't met.
type DeleteFeedbackCategoryRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteFeedbackCategoryRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteFeedbackCategoryRequestMultiError) AllErrors() []error { return m }

// DeleteFeedbackCategoryRequestValidationError is the validation error
// returned by DeleteFeedbackCategoryRequest.Validate if the designated
// constraints aren't met.
type DeleteFeedbackCategoryRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteFeedbackCategoryRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteFeedbackCategoryRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteFeedbackCategoryRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteFeedbackCategoryRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteFeedbackCategoryRequestValidationError) ErrorName() string {
	return "DeleteFeedbackCategoryRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteFeedbackCategoryRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteFeedbackCategoryRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteFeedbackCategoryRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteFeedbackCategoryRequestValidationError{}

// Validate checks the field values on DeleteFeedbackCategoryReply with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteFeedbackCategoryReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteFeedbackCategoryReply with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteFeedbackCategoryReplyMultiError, or nil if none found.
func (m *DeleteFeedbackCategoryReply) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteFeedbackCategoryReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return DeleteFeedbackCategoryReplyMultiError(errors)
	}

	return nil
}

// DeleteFeedbackCategoryReplyMultiError is an error wrapping multiple
// validation errors returned by DeleteFeedbackCategoryReply.ValidateAll() if
// the designated constraints aren't met.
type DeleteFeedbackCategoryReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteFeedbackCategoryReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteFeedbackCategoryReplyMultiError) AllErrors() []error { return m }

// DeleteFeedbackCategoryReplyValidationError is the validation error returned
// by DeleteFeedbackCategoryReply.Validate if the designated constraints
// aren't met.
type DeleteFeedbackCategoryReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteFeedbackCategoryReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteFeedbackCategoryReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteFeedbackCategoryReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteFeedbackCategoryReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteFeedbackCategoryReplyValidationError) ErrorName() string {
	return "DeleteFeedbackCategoryReplyValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteFeedbackCategoryReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteFeedbackCategoryReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteFeedbackCategoryReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteFeedbackCategoryReplyValidationError{}

// Validate checks the field values on
// ListFeedbackCategoryReply_FeedbackCategory with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ListFeedbackCategoryReply_FeedbackCategory) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on
// ListFeedbackCategoryReply_FeedbackCategory with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in
// ListFeedbackCategoryReply_FeedbackCategoryMultiError, or nil if none found.
func (m *ListFeedbackCategoryReply_FeedbackCategory) ValidateAll() error {
	return m.validate(true)
}

func (m *ListFeedbackCategoryReply_FeedbackCategory) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Name

	// no validation rules for CreatedAt

	if len(errors) > 0 {
		return ListFeedbackCategoryReply_FeedbackCategoryMultiError(errors)
	}

	return nil
}

// ListFeedbackCategoryReply_FeedbackCategoryMultiError is an error wrapping
// multiple validation errors returned by
// ListFeedbackCategoryReply_FeedbackCategory.ValidateAll() if the designated
// constraints aren't met.
type ListFeedbackCategoryReply_FeedbackCategoryMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListFeedbackCategoryReply_FeedbackCategoryMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListFeedbackCategoryReply_FeedbackCategoryMultiError) AllErrors() []error { return m }

// ListFeedbackCategoryReply_FeedbackCategoryValidationError is the validation
// error returned by ListFeedbackCategoryReply_FeedbackCategory.Validate if
// the designated constraints aren't met.
type ListFeedbackCategoryReply_FeedbackCategoryValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListFeedbackCategoryReply_FeedbackCategoryValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListFeedbackCategoryReply_FeedbackCategoryValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListFeedbackCategoryReply_FeedbackCategoryValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListFeedbackCategoryReply_FeedbackCategoryValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListFeedbackCategoryReply_FeedbackCategoryValidationError) ErrorName() string {
	return "ListFeedbackCategoryReply_FeedbackCategoryValidationError"
}

// Error satisfies the builtin error interface
func (e ListFeedbackCategoryReply_FeedbackCategoryValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListFeedbackCategoryReply_FeedbackCategory.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListFeedbackCategoryReply_FeedbackCategoryValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListFeedbackCategoryReply_FeedbackCategoryValidationError{}
